manifests/adopt.pp

```diff
@@ -0,0 +1,371 @@
+# Optional forced package types: rpm, rpmbin, tar.gz
+#
+# Variables
+# [*release_major*]
+# Major version release number for java. Used to construct download URL.
+#
+# [*release_minor*]
+# Minor version release number for java. Used to construct download URL.
+#
+# [*install_path*]
+# Base install path for specified version of java. Used to determine if java
+# has already been installed.
+#
+# [*os*]
+# java OS type.
+#
+# [*destination*]
+# Destination directory to save java installer to.  Usually /tmp on Linux and
+# C:\TEMP on Windows.
+#
+# [*creates_path*]
+# Fully qualified path to java after it is installed. Used to determine if
+# java is already installed.
+#
+# [*arch*]
+# java architecture type.
+#
+# [*package_name*]
+# Name of the java installation package to download from github.
+#
+# [*install_command*]
+# Installation command used to install java. Installation commands
+# differ by package_type. 'bin' types are installed via shell command. 'rpmbin'
+# types have the rpms extracted and then forcibly installed. 'rpm' types are
+# forcibly installed.
+#
+# [*spacer*]
+# Spacer to be used in github download url. In major version 8 this is a simple dash
+# in later versions they use a crazy plus sign, which needs to be used in urlencoded
+# format
+#
+# [*download_folder_prefix*]
+# Download folder name begins differently depending on the release. After major release
+# 8, they have given it a dash. Be aware that even if you want to have a JRE, the folder
+# still begins with "jdk"
+#
+# [*release_minor_url*]
+# filled depending on major release. Until major release 8 the minor part needs to be given
+# with a 'b' for build, in later versions it is a underscore or a plus sign, which needs
+# to be stripped for the download url and is replaced with the given spaces (see above)
+#
+# [*_package_type*]
+# Helper variable which gets filled depending on parameter package_type
+#
+# [*_basedir*]
+# Helper variable which gets filled depending on parameter basedir
+#
+# [*_version*]
+# Helper variable which gets filled depending on parameter version
+#
+# [*_version_int*]
+# Helper variable which gets the value of $_version converted to integer
+#
+# [*_append_jre*]
+# Helper variable which gets filled with the string "-jre" if jre was selected to build the correct install path
+#
+# [*_release_minor_package_name*]
+# Helper variable which gets filled with the right minor string depending on the major version
+#
+define java::adopt (
+  $ensure         = 'present',
+  $version        = '8',
+  $version_major  = undef,
+  $version_minor  = undef,
+  $java           = 'jdk',
+  $proxy_server   = undef,
+  $proxy_type     = undef,
+  $basedir        = undef,
+  $manage_basedir = true,
+  $package_type   = undef,
+) {
+
+  # archive module is used to download the java package
+  include ::archive
+
+  # validate java Standard Edition to download
+  if $java !~ /(jre|jdk)/ {
```

Comment:
You can work  with types to avoid this test eg. Enum['jdk', 'jre'] = 'jdk' on define parameters declaration 
https://puppet.com/docs/puppet/6.6/lang_data_type.html